File: Java/javaperformancetester/src/main/java/edu/imepac/javaperformancetester/controllers/AlgoritmosController.java
package edu.imepac.javaperformancetester.controllers;

import edu.imepac.javaperformancetester.dtos.ResponseDto;
import org.springframework.http.HttpStatus;
import org.springframework.http.HttpStatusCode;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.lang.management.ManagementFactory;
import java.lang.management.ThreadMXBean;
import java.math.BigInteger;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

@RestController
@RequestMapping("/algoritmos")
public class AlgoritmosController {

    @GetMapping("/fatorial/{fatorando}")
    public ResponseEntity<ResponseDto<BigInteger>> fatorialRecursivo(@PathVariable int fatorando) {
        ThreadMXBean threadMXBean = ManagementFactory.getThreadMXBean();
        boolean cpuTimeSupported = threadMXBean.isCurrentThreadCpuTimeSupported();
        long cpuStart = cpuTimeSupported ? threadMXBean.getCurrentThreadCpuTime() : 0;

        BigInteger resultado = fatorial(fatorando);

        long cpuEnd = cpuTimeSupported ? threadMXBean.getCurrentThreadCpuTime() : 0;
        long cpuTimeMs = cpuTimeSupported ? (cpuEnd - cpuStart) / 1_000_000 : -1;

        ResponseDto<BigInteger> resposta = new ResponseDto<>(cpuTimeMs, resultado);
        return ResponseEntity.status(HttpStatus.OK).body(resposta);
    }

    @GetMapping("/primos/{limite}")
    public ResponseEntity<ResponseDto<List<Integer>>> primos(@PathVariable int limite) {
        ThreadMXBean threadMXBean = ManagementFactory.getThreadMXBean();
        boolean cpuTimeSupported = threadMXBean.isCurrentThreadCpuTimeSupported();
        long cpuStart = cpuTimeSupported ? threadMXBean.getCurrentThreadCpuTime() : 0;

        List<Integer> resultado = calcularPrimos(limite);

        long cpuEnd = cpuTimeSupported ? threadMXBean.getCurrentThreadCpuTime() : 0;
        long cpuTimeMs = cpuTimeSupported ? (cpuEnd - cpuStart) / 1_000_000 : -1;

        ResponseDto<List<Integer>> resposta = new ResponseDto<>(cpuTimeMs, resultado);
        return ResponseEntity.status(HttpStatus.OK).body(resposta);
    }


    public static BigInteger fatorial(int n) {
        if (n <= 1) return BigInteger.ONE;
        return BigInteger.valueOf(n).multiply(fatorial(n-1));
    }

    public List<Integer> calcularPrimos(int limite) {
        boolean[] ehPrimo = new boolean[limite + 1];
        Arrays.fill(ehPrimo, true);
        ehPrimo[0] = ehPrimo[1] = false;

        for (int i = 2; i * i <= limite; i++) {
            if (ehPrimo[i]) {
                for (int j = i * i; j <= limite; j += i) {
                    ehPrimo[j] = false;
                }
            }
        }

        List<Integer> primos = new ArrayList<>();
        for (int i = 2; i <= limite; i++) {
            if (ehPrimo[i]) primos.add(i);
        }

        return primos;
    }

}

File: Kotlin/kotlinperformancetester/src/main/kotlin/edu/imepac/kotlinperformancetester/controllers/AlgoritmosController.kt
package edu.imepac.javaperformancetester.controllers

import edu.imepac.javaperformancetester.dtos.ResponseDto
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import java.lang.management.ManagementFactory
import java.lang.management.ThreadMXBean
import java.math.BigInteger

@RestController
@RequestMapping("/algoritmos")
class AlgoritmosController {

    @GetMapping("/fatorial/{fatorando}")
    fun fatorialRecursivo(@PathVariable fatorando: Int): ResponseEntity<ResponseDto<BigInteger>> {
        val threadMXBean: ThreadMXBean = ManagementFactory.getThreadMXBean()
        val cpuTimeSupported = threadMXBean.isCurrentThreadCpuTimeSupported
        val cpuStart = if (cpuTimeSupported) threadMXBean.currentThreadCpuTime else 0

        val resultado = fatorial(fatorando)

        val cpuEnd = if (cpuTimeSupported) threadMXBean.currentThreadCpuTime else 0
        val cpuTimeMs = if (cpuTimeSupported) (cpuEnd - cpuStart) / 1_000_000 else -1

        val resposta = ResponseDto(cpuTimeMs, resultado)
        return ResponseEntity.status(HttpStatus.OK).body(resposta)
    }

    @GetMapping("/primos/{limite}")
    fun primos(@PathVariable limite: Int): ResponseEntity<ResponseDto<List<Int>>> {
        val threadMXBean: ThreadMXBean = ManagementFactory.getThreadMXBean()
        val cpuTimeSupported = threadMXBean.isCurrentThreadCpuTimeSupported
        val cpuStart = if (cpuTimeSupported) threadMXBean.currentThreadCpuTime else 0

        val resultado = calcularPrimos(limite)

        val cpuEnd = if (cpuTimeSupported) threadMXBean.currentThreadCpuTime else 0
        val cpuTimeMs = if (cpuTimeSupported) (cpuEnd - cpuStart) / 1_000_000 else -1

        val resposta = ResponseDto(cpuTimeMs, resultado)
        return ResponseEntity.status(HttpStatus.OK).body(resposta)
    }

    fun fatorial(n: Int): BigInteger {
        return if (n <= 1) BigInteger.ONE else BigInteger.valueOf(n.toLong()).multiply(fatorial(n - 1))
    }

    fun calcularPrimos(limite: Int): List<Int> {
        val ehPrimo = BooleanArray(limite + 1) { true }
        if (limite >= 0) ehPrimo[0] = false
        if (limite >= 1) ehPrimo[1] = false

        for (i in 2..Math.sqrt(limite.toDouble()).toInt()) {
            if (ehPrimo[i]) {
                for (j in i * i..limite step i) {
                    ehPrimo[j] = false
                }
            }
        }

        return (2..limite).filter { ehPrimo[it] }
    }
}
